### PR TITLE
Added column!() macro

### PR DESCRIPTION
**Pull request by mvvsmk**
Fixes issue #979 
1) Added the column!() macro using the LOCATION_COLUMN() from gcc_linemap
2) To-Do: add relevant test cases.

Signed-off-by : M V V S Manoj Kumar <mvvsmanojkumar@gmail.com>

The test case I added always fails, I can't figure out whether there is a problem in my test case or there is something wrong with my implementation of the column!() macro. Do let me know where I am going wrong and also if I missed something . :)

### CODE REVIEW

**Review by CohenArthur — APPROVED** [review]
Nice job!!!! The issue is with the typechecking of the expanded fragment. I'll look into why this is not working properly. I think we should leave this PR open in the meantime. Thanks a lot!

**Comment by philberty** [comment]
I think this is a case where our HIR/AST dumps being improved would really help debug this feel free to poke me to look into it, if its a type resolution issue

File: gcc/testsuite/rust/execute/torture/builtin_macros3.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+extern "C" {
+    fn printf(fmt: *const i8, ...);
+}
+
+fn print(s: u32) {
+    printf("%u\n\0" as *const str as *const i8, s);
+}
+
+fn main() -> i32 {
+    let c0 = column!();
```
Comment (philberty) [inline gcc/testsuite/rust/execute/torture/builtin_macros3.rs:15]: I did some testing on this and I think the issues must be relating to transcribing of macros. If you change this line to:

```
let c0:u32 = column!();
```

You will hit the issue the location info here is not the best but the error is:

```
test.rs:15:5: error: expected [u32] got [()]
   15 |     let c0: u32 = column!();
      |     ^       ~
```

The transcribed version of the code is actually

```
let c0:u32 = {}
```

This is why we get the errors we see here, I think @CohenArthur will have a better idea what to look at next

**Comment by CohenArthur** [comment]
@mvvsmk I figured it out :) The builtin transcriber would never get called because you hadn't registered it

```diff
diff --git a/gcc/rust/util/rust-hir-map.cc b/gcc/rust/util/rust-hir-map.cc
index 9190bd92a43..5b0417b4549 100644
--- a/gcc/rust/util/rust-hir-map.cc
+++ b/gcc/rust/util/rust-hir-map.cc
@@ -750,6 +750,7 @@ Mappings::insert_macro_def (AST::MacroRulesDefinition *macro)
     builtin_macros = {
       {"assert", MacroBuiltin::assert},
       {"file", MacroBuiltin::file},
+      {"column", MacroBuiltin::column},
     };
 
   auto builtin = builtin_macros.find (macro->get_rule_name ());
```

I think this makes it clear that adding a builtin is not obvious enough. I'll add some documentation to the rust-macro-builtins header

**Comment by mvvsmk** [comment]
Thank you  @CohenArthur @philberty for helping me out. I got an Idea of what to do now, that is register the buitltin macro in the Mappings class, as these work diffrently as builtin transcribers are responsible for handling all match arms of the macro.

**Review by CohenArthur — APPROVED** [review]
LGTM!

**Comment by CohenArthur** [comment]
bors: cancel

**Comment by CohenArthur** [comment]
Sorry @mvvsmk one last thing, could you rebase all your commits into one? This way the buildbot about tests not passing in [03bbcf1](https://github.com/Rust-GCC/gccrs/pull/1004/commits/03bbcf17207570c7659ef9933b5f89afebc4ff3b). Thanks!

**Comment by mvvsmk** [comment]
Sure!

**Comment by CohenArthur** [comment]
Perfect thanks a lot! Great job!

bors r+

**Pull request merged**